native/dgesvd.go

```diff
@@ -0,0 +1,1241 @@
+// not used.
+//
+// work is a slice for storing temporary memory, and lwork is the usable size of
+// the slice. lwork must be at least max(5*min(m,n), 3*min(m,n)+max(m,n)).
+// If lwork == -1, instead of performing Dgesvd, the optimal work length will be
+// stored into work[0]. Dgesvd will panic if the working memory has insufficient
+// storage.
+//
+// Dgesvd returns whether the decomposition successfully completed.
+func (impl Implementation) Dgesvd(jobU, jobVT lapack.SVDJob, m, n int, a []float64, lda int, s, u []float64, ldu int, vt []float64, ldvt int, work []float64, lwork int) (ok bool) {
+	checkMatrix(m, n, a, lda)
+	if jobU == lapack.SVDAll {
+		checkMatrix(m, m, u, ldu)
+	} else if jobU == lapack.SVDInPlace {
+		checkMatrix(m, min(m, n), u, ldu)
+	}
+	if jobVT == lapack.SVDAll {
+		checkMatrix(n, n, vt, ldvt)
+	} else if jobVT == lapack.SVDInPlace {
+		checkMatrix(min(m, n), n, vt, ldvt)
+	}
+	if jobU == lapack.SVDOverwrite && jobVT == lapack.SVDOverwrite {
+		panic("lapack: both jobU and jobVT are lapack.SVDOverwrite")
+	}
+	if len(s) < min(m, n) {
+		panic(badS)
+	}
+	if jobU == lapack.SVDOverwrite || jobVT == lapack.SVDOverwrite {
+		panic(noSVDO)
+	}
+	minWork := max(5*min(m, n), 3*min(m, n)+max(m, n))
+	if lwork != -1 {
+		if len(work) < lwork {
+			panic(badWork)
+		}
+		if lwork < minWork {
+			panic(badWork)
+		}
+	}
+	if m == 0 || n == 0 {
+		return true
+	}
+
+	wantua := jobU == lapack.SVDAll
+	wantus := jobU == lapack.SVDInPlace
+	wantuas := wantua || wantus
+	wantuo := jobU == lapack.SVDOverwrite
+	wantun := jobU == lapack.None
+
+	wantva := jobVT == lapack.SVDAll
+	wantvs := jobVT == lapack.SVDInPlace
+	wantvas := wantva || wantvs
+	wantvo := jobVT == lapack.SVDOverwrite
+	wantvn := jobVT == lapack.None
+
+	bi := blas64.Implementation()
+	var mnthr int
+
+	// The netlib implementation checks has this at only length 1. Our implementation
+	// checks all input sizes before examining the l == -1 case.
+	dum := make([]float64, m*n)
+
+	// Compute optimal space for subroutines.
+	maxwrk := 1
+	opts := string(jobU) + string(jobVT)
+	var wrkbl, bdspac int
+	if m >= n {
+		mnthr = impl.Ilaenv(6, "DGESVD", opts, m, n, 0, 0)
+		bdspac = 5 * n
+		impl.Dgeqrf(m, n, a, lda, dum, dum, -1)
+		lwork_dgeqrf := int(dum[0])
+		impl.Dorgqr(m, n, n, a, lda, dum, dum, -1)
+		lwork_dorgqr_n := int(dum[0])
+		impl.Dorgqr(m, m, n, a, lda, dum, dum, -1)
+		lwork_dorgqr_m := int(dum[0])
+		impl.Dgebrd(n, n, a, lda, s, dum, dum, dum, dum, -1)
+		lwork_dgebrd := int(dum[0])
+		impl.Dorgbr(lapack.ApplyP, n, n, n, a, lda, dum, dum, -1)
+		lwork_dorgbr_p := int(dum[0])
+		impl.Dorgbr(lapack.ApplyQ, n, n, n, a, lda, dum, dum, -1)
+		lwork_dorgbr_q := int(dum[0])
+
+		if m >= mnthr {
+			// m >> n
+			if wantun {
+				// Path 1
+				maxwrk = n + lwork_dgeqrf
+				maxwrk = max(maxwrk, 3*n+lwork_dgebrd)
+				if wantvo || wantvas {
+					maxwrk = max(maxwrk, 3*n+lwork_dorgbr_p)
+				}
+				maxwrk = max(maxwrk, bdspac)
```

Comment:
Yea. The function comment specifically lists the actual min work, so it's not clear why it's needed.